frcxx_nft/src/lib.rs

```diff
@@ -272,26 +349,48 @@ where
     /// Transfers a token that the caller is an operator for
     pub fn transfer_from(
         &mut self,
+        owner: &Address,
         operator: &Address,
         recipient: &Address,
         token_ids: &[TokenID],
         operator_data: RawBytes,
         token_data: RawBytes,
-    ) -> Result<ReceiverHook<TransferFromIntermediate>> {
+    ) -> Result<ReceiverHook<TransferIntermediate>> {
         // Attempt to instantiate the accounts if they don't exist
-        let operator = self.msg.resolve_or_init(operator)?;
-        let recipient = self.msg.resolve_or_init(recipient)?;
+        let owner_id = self.msg.resolve_id(owner)?;
+        let operator_id = self.msg.resolve_id(operator)?;
+        let recipient_id = self.msg.resolve_or_init(recipient)?;
 
-        let hook = self.state.operator_transfer_tokens(
-            &self.bs,
-            operator,
-            recipient,
-            token_ids,
+        let intermediate = self.transaction(|state, bs| {
+            let owner_map = state.get_owner_data_hamt(bs)?;
+            let account_operator =
+                NFTState::is_account_operator(&owner_map, owner_id, operator_id)?;
+            let intermediate = state.transfer(
+                bs,
+                token_ids,
+                owner_id,
+                recipient_id,
+                &|token_data, token_id| {
+                    NFTState::assert_owns_token(token_data, token_id, owner_id)?;
+                    if !account_operator {
+                        NFTState::assert_token_level_approval(token_data, token_id, operator_id)
+                    } else {
+                        Ok(())
+                    }
```

Comment:
There's a small bit of repeated logic here from transfer. If it's easy to factor out, let's do so, but no biggie if it's not